src/LCTrackClusterAssociation/TrackClusterAssociationAlgorithm.cc

```diff
@@ -113,81 +112,100 @@ StatusCode TrackClusterAssociationAlgorithm::Run()
+                            // cache all hits that are nearby the track
+                            tracks_to_hits.emplace(hash_key,hit.data);
+                            // add to the list of nearby clusters
+                            nearby_clusters.insert(assc_cluster->second);
+                        }
                     }
+                    found_hits.clear();
                 }
-                found_hits.clear();
             }
-        }
 
-        ClusterList nearbyClusterList(nearby_clusters.begin(), nearby_clusters.end());
-        nearbyClusterList.sort(SortingHelper::SortClustersByNHits);
-        nearby_clusters.clear();
+            ClusterList nearbyClusterList(nearby_clusters.begin(), nearby_clusters.end());
+            nearbyClusterList.sort(SortingHelper::SortClustersByNHits);
+            nearby_clusters.clear();
 
-        // Identify the closest cluster and also the closest cluster below a specified hadronic energy threshold
-        for (const Cluster *const pCluster : nearbyClusterList)
-        {
-            if (0 == pCluster->GetNCaloHits())
-                continue;
-
-            float trackClusterDistance(std::numeric_limits<float>::max());
-            if (STATUS_CODE_SUCCESS != lc_content::ClusterHelper::GetTrackClusterDistance(pTrack, pCluster, m_maxSearchLayer, m_parallelDistanceCut,
-                m_minTrackClusterCosAngle, trackClusterDistance))
+            // Identify the closest cluster and also the closest cluster below a specified hadronic energy threshold
+            for (const Cluster *const pCluster : nearbyClusterList)
             {
-                continue;
-            }
+                if (0 == pCluster->GetNCaloHits())
+                    continue;
 
-            const float energyDifference(std::fabs(pCluster->GetHadronicEnergy() - pTrack->GetEnergyAtDca()));
+                float trackClusterDistance(std::numeric_limits<float>::max());
+                if (STATUS_CODE_SUCCESS != lc_content::ClusterHelper::GetTrackClusterDistance(trackState, pCluster, m_maxSearchLayer, m_parallelDistanceCut,
+                    m_minTrackClusterCosAngle, trackClusterDistance))
+                {
+                    continue;
+                }
 
-            if (pCluster->GetHadronicEnergy() > m_lowEnergyCut)
-            {
-                if ((trackClusterDistance < minDistance) || ((trackClusterDistance == minDistance) && (energyDifference < minEnergyDifference)))
+                const float energyDifference(std::fabs(pCluster->GetHadronicEnergy() - pTrack->GetEnergyAtDca()));
+
+                if (pCluster->GetHadronicEnergy() > m_lowEnergyCut)
                 {
-                    minDistance = trackClusterDistance;
-                    pBestCluster = pCluster;
-                    minEnergyDifference = energyDifference;
+                    if ((trackClusterDistance < minDistance) || ((trackClusterDistance == minDistance) && (energyDifference < minEnergyDifference)))
```

Comment:
The reason for this nonsense? Because we once had issues with the ordering of the Clusters coming out of an unordered container. Reproducibility is obviously very important and something we are completely on top of now (but there are code relics dating from when we had low-level exposure to unordered containers and tried to work around the issues).